### PR TITLE
feat: allow passing gradient to .backward() to compute vjp

### PR DESCRIPTION
**Pull request by davidgonmar**
This allows the user to pass a `gradient` parameter to Tensor.backward. The API is similar to the one in PyTorch: if the tensor is not a scalar, then gradient must be passed. If it is scalar and it is not passed, it is assumed to be 1.0.
This means that backward can generalize to vector-jacobian products (can be seen as a 'generalization' of the gradient).

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -724,22 +724,22 @@ def _walk(node, visited):
...
     ```
     """
-    assert self.shape == tuple(), f"backward can only be called for scalar tensors, but it has shape {self.shape})"
-
+    assert self.shape == tuple() or gradient is not None, "backward must be called on a scalar tensor or with the gradient argument"
+    assert gradient is None or gradient.shape == self.shape, f"gradient shape must match tensor shape, {gradient.shape!r} != {self.shape!r}"
     # fill in the first grad with one. don't use Tensor.ones because we don't need contiguous
     # this is "implicit gradient creation"
-    self.grad = Tensor(1.0, dtype=self.dtype, device=self.device, requires_grad=False)
+    self.grad = Tensor(1.0, dtype=self.dtype, device=self.device, requires_grad=False) if gradient is None else gradient
```
Comment (geohot) [inline tinygrad/tensor.py:742]: Create the implicit gradient before the asserts

Comment (davidgonmar) [reply tinygrad/tensor.py:742]: you mean something like
```python
self.grad = Tensor(1.0, dtype=self.dtype, device=self.device, requires_grad=False)
assert self.shape == tuple() or gradient is not None, "backward must be called on a scalar tensor or with the gradient argument"
assert gradient is None or gradient.shape == self.shape, f"gradient shape must match tensor shape, {gradient.shape!r} != {self.shape!r}"
self.grad = self.grad if gradient is None else gradient
```
?

Comment (davidgonmar) [reply tinygrad/tensor.py:742]: but this could be incorrect. 'gradient' (which is really a vector jacobian product) must be of the same shape as the tensor (self.shape == self.grad.shape) always. So creating an 'implicit' (scalar with value 1) grad (vector-jacobian product) for a tensor of different shape is incorrect. Did you mean something else?

Comment (geohot) [reply tinygrad/tensor.py:742]: No. The same code.

```
if gradient is None:
  assert self.shape == tuple()
  gradient = ...```

Comment (davidgonmar) [reply tinygrad/tensor.py]: > No. The same function, don't change self.grad
> 
> ```
> if gradient is None:
>   assert self.shape == tuple()
>   gradient = ...```
> ```

Corrected in the last commit
